modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/function/MBFunction_notEqualTo.java

```diff
@@ -0,0 +1,101 @@
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package org.geotools.mbstyle.function;
+
+import static org.geotools.filter.capability.FunctionNameImpl.parameter;
+
+import org.geotools.filter.FunctionExpressionImpl;
+import org.geotools.filter.capability.FunctionNameImpl;
+import org.opengis.filter.capability.FunctionName;
+
+/**
+ * MapBox Expression function that returns {@link java.lang.Boolean#FALSE} if two expressions are equivalent,
+ * {@link java.lang.Boolean#TRUE} otherwise. It is slightly different from the GeoTools "notEqualTo" function as it
+ * treats NULLs as equal (instead of not equal), and it does not compare Object.toString() values, which would result in
+ * false equivalences for things like Boolean.TRUE compared to the literal string "true".
+ * <p>
+ * Format:
+ * </p>
+ * <pre>
+ *     ["mbNotEqualTo", &lt;expression&gt;, &lt;expression&gt;]
+ * </pre>
+ * <p>
+ * Examples:
+ * </p>
+ * <p>
+ * <table border="1" cellpadding="3">
+ *   <tr>
+ *     <th align="center">Expression</th>
+ *     <th align="center">Output</th>
+ *   </tr>
+ *   <tr>
+ *     <td>["mbNotEqualTo", "aString", "aString"]</td>
+ *     <td align="center">false</td>
+ *   </tr>
+ *   <tr>
+ *     <td>["mbNotEqualTo", null, null]</td>
+ *     <td align="center">false</td>
+ *   </tr>
+ *   <tr>
+ *     <td>["mbNotEqualTo", true, "true"]</td>
+ *     <td align="center">true</td>
+ *   </tr>
+ *   <tr>
+ *     <td>["mbNotEqualTo", 5, 5.0]</td>
+ *     <td align="center">false</td>
+ *   </tr>
+ * </table>
+ * </p>
+ */
+class MBFunction_notEqualTo extends FunctionExpressionImpl {
+
+    public static FunctionName NAME = new FunctionNameImpl("mbNotEqualTo",
+        parameter("mbNotEqualTo", Boolean.class),
+        parameter("a", Object.class),
+        parameter("b", Object.class));
+
+    MBFunction_notEqualTo() {
+        super(NAME);
+    }
+
+    /**
+     * @see org.geotools.filter.FunctionExpressionImpl#equals(java.lang.Object)
+     */
+    @Override
+    public Object evaluate(Object feature) {
+        Object arg0;
+        Object arg1;
+
+        try { // attempt to get value and perform conversion
+            arg0 = (Object) getExpression(0).evaluate(feature);
+        } catch (Exception e) // probably a type error
+        {
+            throw new IllegalArgumentException(
+                    "Filter Function problem for function equalTo argument #0 - expected type Object");
+        }
+
+        try { // attempt to get value and perform conversion
+            arg1 = (Object) getExpression(1).evaluate(feature);
```

Comment:
Logic should attempt to match types using expression.evaulate(feature, arg0.getClass() ), and if that is not available consider fallbacks such as evaluating both to String.

Using evaulate(feature, class) is preferable to using the geotools converters facilities by hand as a given expression may have a better idea how to evaluate to a number, color, or string as required.

Reply:
I guess the reason I did it this way is that you can't really tell what Class type the expression will evaluate to in some cases. And I can't use a String fallback because in MapBox world, `"true"` is not equal to `Boolean.TRUE`. Doing a String comparison of them would result in a false positive for equalTo (or false negative for notEqualTo).

Since both these functions use the same logic (one is just the "not" of the other), I was able to keep the logic in one place (https://github.com/geotools/geotools/pull/1865/files#diff-7d33831b31b1087f88bd7b4f497dd00bR34). That utility method handles the class type comparisons and relies on the `equals()` implementation of the evaluated expressions, which seems to be how the expression works in a MapBox style.